Src/Support/Google.Apis.Auth/OAuth2/AwsExternalAccountCredential.AwsSignedSubjectToken.cs

```diff
@@ -0,0 +1,279 @@
+    {
+        /// <summary>
+        /// Partial implementation of AWS signed request, enough to support signing
+        /// of a, usually, AWS GetCallerIdentity request. The signed request is sent
+        /// to Google's subject token service as the subject token to be exchanged for
+        /// access tokens. Google STS triggers the request as specified by the signed
+        /// request to verify the callers identity.
+        /// </summary>
+        private sealed class AwsSignedSubjetToken
+        {
+            private const string StVerificationHttpMethod = "POST";
+            private const string BasicFormatIso8601 = "yyyyMMddTHHmmssZ";
+            private const string ShortDateFormat = "yyyyMMdd";
+            private const string AwsRequestType = "aws4_request";
+            private const string AwsSha256Designation = "AWS4-HMAC-SHA256";
+
+            [JsonProperty("url")]
+            public string Url { get; }
+
+            [JsonProperty("method")]
+            public string HttpMethod { get; }
+
+            [JsonProperty("body")]
+            public string Body { get; }
+
+            [JsonProperty("headers")]
+            public AwsSignedSubjectTokenHeader[] Headers { get; }
+
+            private AwsSignedSubjetToken(string url, string httpMethod, string body, AwsSignedSubjectTokenHeader[] headers)
+            {
+                Url = url;
+                HttpMethod = httpMethod;
+                Body = body;
+                Headers = headers;
+            }
+
+            public sealed class AwsSignedSubjectTokenHeader
+            {
+                [JsonProperty("key")]
+                public string Key { get; }
+
+                [JsonProperty("value")]
+                public string Value { get; }
+
+                internal AwsSignedSubjectTokenHeader(string key, string value)
+                {
+                    Key = key;
+                    Value = value;
+                }
+            }
+            internal static AwsSignedSubjetToken Create(AwsSecurityCredentials credentials, AwsRegion region, Uri verificationRequestUrl, string audience, IClock clock)
+            {
+                credentials.ThrowIfNull(nameof(credentials));
+                region.ThrowIfNull(nameof(region));
+                verificationRequestUrl.ThrowIfNull(nameof(verificationRequestUrl));
+                audience.ThrowIfNull(nameof(audience));
+                clock.ThrowIfNull(nameof(clock));
+
+                const string newLine = "\n";
+
+                // https://docs.aws.amazon.com/general/latest/gr/sigv4_signing.html
+
+                DateTime utcNow = clock.UtcNow;
+                string host = verificationRequestUrl.Host;
+                // The service name is the first part of the host, for instance
+                // iam in iam.amazonaws.com or sts in sts.amazonaws.com
+                string serviceName = host.Split(new char[] { '.' }, StringSplitOptions.RemoveEmptyEntries)[0];
```

Comment:
Will this throw an IndexOutOfRangeException if Host is empty? (Given that you're removing empty entries.) Possibly just don't remove empty entries, so there's at least one result?